src/librustc_lint/builtin.rs

```diff
@@ -1307,3 +1307,36 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
         }
     }
 }
+
+/// Lint for items marked `pub` that aren't exported to other crates
+pub struct UnexportedPub;
+
+declare_lint! {
+    UNEXPORTED_PUB,
+    Allow,
+    "`pub` items not reachable from crate root"
+}
+
+impl LintPass for UnexportedPub {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNEXPORTED_PUB)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnexportedPub {
+    fn check_item(&mut self, cx: &LateContext, item: &hir::Item) {
+        if !cx.access_levels.is_exported(item.id) && item.vis == hir::Visibility::Public {
```

Comment:
As I mentioned on gitter, this should use `is_reachable` instead of `is_exported`.
With `is_exported` this lint will give false positives on anything that is leaked through any other means than `use` (type aliases, return types, field types). All these things still need to be `pub` to be usable from other crates.

Reply:
Test to add:
```rust
mod m {
    pub struct S; // The lint should not fire on this struct
}

pub fn f() -> m::S { m::S }
```